numpy_financial/_financial.py

```diff
@@ -840,20 +825,40 @@ def irr(values, *, guess=None, tol=1e-12, maxiter=100, raise_exceptions=False):
     #
     # which we solve using Newton-Raphson and then reverse out the solution
     # as eirr = g - 1 (if we are close enough to a solution)
-    npv_ = np.polynomial.Polynomial(values[::-1])
-    d_npv = npv_.deriv()
-    g = 1 + guess
+    
+    g = np.roots(values)
+    IRR = np.real(g[np.isreal(g)]) - 1
 
-    for _ in range(maxiter):
-        delta = npv_(g) / d_npv(g)
-        if abs(delta) < tol:
-            return g - 1
-        g -= delta
+    # realistic IRR
+    IRR = IRR[IRR >= -1]
 
-    if raise_exceptions:
-        raise IterationsExceededError('Maximum number of iterations exceeded.')
+    # if no real solution
+    if len(IRR) == 0:
+        if raise_exceptions:
+            raise NoRealSolutionError("No real solution is found for IRR.")
+        return np.nan
 
-    return np.nan
+    # if only one real solution
+    if len(IRR) == 1:
+        return IRR[0]
+    
+    # below is for the situation when there are more than 2 real solutions.
+    # check sign of all IRR solutions
+    same_sign = np.all(IRR > 0) if IRR[0] > 0 else np.all(IRR < 0)
+    
+    # if the signs of IRR solutions are not the same, first filter potential IRR
+    # by comparing the total positive and negative cash flows.
+    if not same_sign:
+        pos = sum(values[values>0])
+        neg = sum(values[values<0])
+        if pos > neg:
```

Comment:
I don't understand _why_ you are trying to filter the by the total cashflow. Is this just some heuristic you have picked or is there a more formal process going on here?

Reply:
This is just some heuristic I have picked. It works when there are real roots with different signs.

I can give you a simple example and walk you through it:

real roots = -0.5 and 0.1
net sum of the cash flows = +10000

As the net sum > 0, we can kind of exclude the possibility that IRR is -0.5 and keep only 0.1.

Reply:
That makes sense. Thanks for clarifying.